CompilerSource/general/bettersystem.cpp

```diff
@@ -252,20 +256,90 @@ void myReplace(std::string& str, const std::string& oldStr, const std::string& n
       return e_exec(cmd, eCenviron);
     }
 #else
+    #include <vector>
+    #include <algorithm>
+    #include <cstdlib>
+    #include <csignal>
+
     #include <fcntl.h>
     #include <unistd.h>
     #include <sys/wait.h>
     #include <sys/stat.h>
+    #include <sys/types.h>
 
     extern char **environ;
     const mode_t laxpermissions = S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH;
 
-#if CURRENT_PLATFORM_ID ==  OS_MACOSX
+#if CURRENT_PLATFORM_ID == OS_MACOSX
+    #include <libproc.h>
+    #include <sys/proc_info.h>
     #include <crt_externs.h>
     #define environ (*_NSGetEnviron())
     extern char **environ;
 #endif
 
+    using std::vector;
+
+#if CURRENT_PLATFORM_ID == OS_FREEBSD
+    void ProcIdFromParentProcId(pid_t parentProcId, pid_t **procId, int *size) {
+      vector<pid_t> vec; int i = 0; int cntp;
+      if (kinfo_proc *proc_info = kinfo_getallproc(&cntp)) {
+        for (int j = 0; j < cntp; j++) {
+          if (proc_info[j].ki_ppid == parentProcId) {
+            vec.push_back(proc_info[j].ki_pid); i++;
+          }
+        }
+        free(proc_info);
+      }
+      *procId = (pid_t *)malloc(sizeof(pid_t) * vec.size());
+      if (procId) {
+        std::copy(vec.begin(), vec.end(), *procId);
+        *size = i;
+      }
+    }
+#elif CURRENT_PLATFORM_ID == OS_MACOSX
+    void ParentProcIdFromProcId(pid_t procId, pid_t *parentProcId) {
+      proc_bsdinfo proc_info;
+      if (proc_pidinfo(procId, PROC_PIDTBSDINFO, 0, &proc_info, sizeof(proc_info)) > 0) {
+        *parentProcId = proc_info.pbi_ppid;
+      }
+    }
+
+    void ProcIdFromParentProcId(pid_t parentProcId, pid_t **procId, int *size) {
+      vector<pid_t> vec; int i = 0;
+      int cntp = proc_listpids(PROC_ALL_PIDS, 0, nullptr, 0);
+      vector<pid_t> proc_info(cntp);
+      std::fill(proc_info.begin(), proc_info.end(), 0);
+      proc_listpids(PROC_ALL_PIDS, 0, &proc_info[0], sizeof(pid_t) * cntp);
+      for (int j = cntp; j > 0; j--) {
+        if (proc_info[j] == 0) { continue; }
+        pid_t ppid; ParentProcIdFromProcId(proc_info[j], &ppid);
+        if (ppid == parentProcId) {
+          vec.push_back(proc_info[j]); i++;
+        }
+      }
+      *procId = (pid_t *)malloc(sizeof(pid_t) * vec.size());
+      if (procId) {
+        std::copy(vec.begin(), vec.end(), *procId);
+        *size = i;
+      }
+    }
+#endif
```

Comment:
You might as well just add an `#else` and implement `ProcIdFromParentProcId` as `waitpid(-fk,&result,__WALL);`. The `options` int seems to be zero for the other platforms, so you might consider dropping it rather than just ignoring it.

Reply:
Good idea, just did it.